packages/neon_http_client/lib/src/interceptors/http_interceptor.dart

```diff
@@ -0,0 +1,26 @@
+import 'dart:async';
+
+import 'package:http/http.dart' as http;
+
+/// Interceptor that can manipulate http requests and responses.
+abstract interface class HttpInterceptor {
+  /// Whether this interceptor should intercept requests.
+  bool shouldInterceptRequest(http.BaseRequest request);
+
+  /// Intercepts the given [request].
+  ///
+  /// Provided requests are not finalized yet. It is an error for an interceptor
+  /// to request itself.
+  FutureOr<http.BaseRequest> interceptRequest({required http.BaseRequest request});
+
+  /// Whether this interceptor should intercept response.
+  bool shouldInterceptResponse(http.StreamedResponse response);
+
+  /// Intercepts the given [response].
+  ///
+  /// Until package:http 2.0 makes [http.BaseResponseWithUrl] mandatory the request url is used.
+  FutureOr<http.StreamedResponse> interceptResponse({
+    required http.StreamedResponse response,
+    required Uri url,
+  });
```

Comment:
Wouldn't it make more sense to use http.Response here as then the interceptors can also modify the body?

Reply:
https://pub.dev/documentation/http/latest/http/Client/send.html

The send function of the client itself expects a streamed response.
In case the body needs to be modified, a stream transformer is probably the right thing to use